source/encoding.c

```diff
@@ -0,0 +1,316 @@
+static const uint8_t bytes_4 = 4 * 8;
+static const uint8_t bytes_3 = 3 * 8;
+static const uint8_t bytes_2 = 2 * 8;
+static const uint8_t bytes_1 = 1 * 8;
+static const uint8_t mask = 0xff;
+
+
+int aws_hex_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size) {
+
+    size_t encoded_len = (to_encode_len << 1) + 1;
+
+    if(encoded_len <= to_encode_len) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+
+    if (!output) {
+        *output_size = encoded_len;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (*output_size < encoded_len) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t written = 0;
+    for (size_t i = 0; i < to_encode_len; ++i) {
+
+        output[written++] = hex_chars[to_encode[i] >> 4 & 0x0f];
+        output[written++] = hex_chars[to_encode[i] & 0x0f];
+    }
+
+    output[written++] = '\0';
+    *output_size = written;
+
+    return AWS_OP_SUCCESS;
+}
+
+static uint8_t char_to_int(char character) {
+    if (character >= 'a' && character <= 'f') {
+        return (uint8_t)(10 + (character - 'a'));
+    }
+
+    if (character >= 'A' && character <= 'F') {
+        return (uint8_t)(10 + (character - 'A'));
+    }
+
+    return (uint8_t)(character - '0');
+}
+
+int aws_hex_decode(const char *to_decode, size_t to_decode_len, uint8_t *output, size_t *output_size) {
+    size_t decoded_length = (to_decode_len + 1) >> 1;
+
+    if (!output) {
+        *output_size = decoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (*output_size < decoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t written = 0;
+
+    size_t i = 0;
+
+    //if the buffer isn't even, prepend a 0 to the buffer.
+    if (to_decode_len & 0x01) {
+        i = 1;
+        output[written++] = char_to_int(to_decode[i]);
+    }
+
+    for (; i < to_decode_len; i += 2) {
+        uint8_t value = char_to_int(to_decode[i]) << 4;
+        value |= char_to_int(to_decode[i + 1]);
+        output[written++] = value;
+    }
+
+    *output_size = written;
+
+    return AWS_OP_SUCCESS;
+}
+
+static size_t calculate_base64_decoded_length(const char *input, size_t len) {
+    if (len == 0) {
+        return 0;
+    }
+
+    size_t padding = 0;
+
+    if (len >= 2 && input[len - 1] == '=' && input[len - 2] == '=') { /*last two chars are = */
+        padding = 2;
+    }
+    else if (input[len - 1] == '=') { /*last char is = */
+        padding = 1;
+    }
+
+    return (len * 3 / 4 - padding);
+}
+
+static size_t calculate_base64_encoded_length(size_t buffer_len) {
+    return 4 * ((buffer_len + 2) / 3) + 1; /* plus one for the NULL terminator */
+}
+
+int aws_base64_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size) {
+
+    size_t encoded_length = calculate_base64_encoded_length(to_encode_len);
+    if(!output) {
+        *output_size = encoded_length;
+        return AWS_OP_SUCCESS;
```

Comment:
We should probably have clearer semantics around our return value. For example, we could declare that `*output_size` is always set, and `AWS_OP_SUCCESS` is returned only if the input is fully converted to the output.

Or, probably better would be to have a separate function to compute the length required for the output buffer based solely on input length, and just fail if the output buffer is too small here.